### PR TITLE
Project addition/deletion

### PR DESCRIPTION
**Pull request by vladak**
The overall goal is to avoid indexing of all projects (i.e. `OpenGrok index /var/opengrok/src`) in order to discover new and remove deleted projects from the configuration. Thus one can start with no index and no projects and then incrementally add them and index them.

It basically works like this:

0. bootstrap initial configuration:
  `./OpenGrok bootstrap`
   - this will create `/var/opengrok/etc/configuration.xml` with basic set of properties. If more is needed `./tools/Messages -n config -t set "set foo = bar"` can be used
1.  add a new project `foo`:
  `./tools/Messages -t foo -n project add`
2. the project `foo` is now visible in the configuration however is not yet searchable. Store the config in a file so that indexer can see the project and its repositories.
  `./tools/Messages -n config -t getconf > /var/opengrok/etc/configuration.xml`
3. index the project. It will become searchable after that.
  `OPENGROK_READ_XML_CONFIGURATION=/var/opengrok/etc/configuration.xml ./OpenGrok indexpart /var/opengrok/src /foo`
4. make the project "indexed" status of the project persistent so that if webapp is redeployed the project will be still visible:
  `./tools/Messages -n config -t getconf > /var/opengrok/etc/configuration.xml`

If an `add` message is sent that matches existing project, the list of repositories for that project will be refreshed. This is handy when repositories are added/deleted.

### CODE REVIEW

**Comment by tarzanek** [comment]
nice!
I will have a look and test this on Friday
if you want, please merge

**Comment by vladak** [comment]
I'd like someone to take a look at the code first.

**Comment by vladak** [comment]
Part of the changes that introduce some noise are:
  - `runHgCommand()` changed to use elipsis
  -

**Comment by vladak** [comment]
The index timestamp will be updated also for each partial reindex. This means the timestamp will no longer reference all projects' indexes, rather date of any last reindex.

File: src/org/opensolaris/opengrok/history/FileHistoryCache.java [hunk 1]
```diff
@@ -64,10 +64,11 @@
     private static final Logger LOGGER = LoggerFactory.getLogger(FileHistoryCache.class);
 
     private final Object lock = new Object();
-    private String historyCacheDirName = "historycache";
-    private String latestRevFileName = "OpenGroklatestRev";
+    private final String historyCacheDirName = "historycache";
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/history/FileHistoryCache.java]: may be static?

File: src/org/opensolaris/opengrok/history/FileHistoryCache.java [hunk 2]
```diff
@@ -64,10 +64,11 @@
...
+    private final String latestRevFileName = "OpenGroklatestRev";
     private boolean historyIndexDone = false;
-    
+    public final static String historyDirName = "historycache";
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/history/FileHistoryCache.java]: what is the point of having these two?
`private final String historyCacheDirName = "historycache";`
`public final static String historyDirName = "historycache";`

File: src/org/opensolaris/opengrok/history/HistoryGuru.java [hunk 3]
```diff
@@ -635,6 +641,12 @@ public void removeCache(Collection<String> repositories) throws HistoryException
                         new Object[]{r.getDirectoryName(), e.getLocalizedMessage()});
             }
         }
+
+        return repos;
+    }
+
+    public void removeCache(Collection<String> repositories) throws HistoryException {
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/history/HistoryGuru.java]: a brief javadoc would be nice

File: src/org/opensolaris/opengrok/history/HistoryGuru.java [hunk 4]
```diff
@@ -764,7 +776,7 @@ public void invalidateRepositories(Collection<? extends RepositoryInfo> repos) {
             boolean verbose = RuntimeEnvironment.getInstance().isVerbose();
 
             if (verbose) {
-                LOGGER.log(Level.FINE, "invalidating repositories");
+                LOGGER.log(Level.FINE, "invalidating " + repos.size() + " repositories");
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/history/HistoryGuru.java]: using log template is preferred

File: src/org/opensolaris/opengrok/configuration/RuntimeEnvironment.java [hunk 5]
```diff
@@ -1024,6 +1045,14 @@ public void setScopesEnabled(boolean scopesEnabled) {
         threadConfig.get().setScopesEnabled(scopesEnabled);
     }
 
+    public boolean isProjectsEnabled() {
+        return threadConfig.get().isProjectsEnabled();
+    }
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/configuration/RuntimeEnvironment.java]: - shouldn't we use this in Util.dumpConfiguration?
- shouldn't we use this in env.hasProjects?

File: test/org/opensolaris/opengrok/configuration/messages/ProjectMessageTest.java [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+import org.opensolaris.opengrok.index.Indexer;
+import static org.opensolaris.opengrok.util.IOUtils.removeRecursive;
+import org.opensolaris.opengrok.util.TestRepository;
+
+/**
+ *
+ * @author vkotal
+ */
+@ConditionalRun(condition = RepositoryInstalled.MercurialInstalled.class)
+// XXX need svn + git
```
Comment (tulinkry) [inline test/org/opensolaris/opengrok/configuration/messages/ProjectMessageTest.java]: ?

File: test/org/opensolaris/opengrok/configuration/messages/ProjectMessageTest.java [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        m.apply(env);
+        Assert.assertEquals(3, env.getProjects().size());
+        Assert.assertEquals(3, env.getRepositories().size());
+        Assert.assertEquals(3, env.getProjectRepositoriesMap().size());
+
+        // Run the indexer (ala 'indexpart') so that data directory is populated.
+        ArrayList<String> subFiles = new ArrayList<>();
+        subFiles.add("/git");
+        subFiles.add("/mercurial");
+        subFiles.add("/git");
```
Comment (tulinkry) [inline test/org/opensolaris/opengrok/configuration/messages/ProjectMessageTest.java]: git twice?

File: tools/Messages [hunk 8]
```diff
@@ -62,6 +62,14 @@ Usage()
...
     echo "                      Returns text describing the action."
     echo "                      The change is NOT persistent when reindex or redeploy!"
+    echo "    project:"
+    echo "     - project names to be added are specified as <tags>"
+    echo "     - command is specified in message <text>:"
+    echo "       - \"add\" - adds project(s) and its repositories to the configuration."
+    echo "                   If the project already exists, refresh list of its repositories."
+    echo "       - \"delete\" - removes project(s) and its repositores from the configuration"
+    echo "                      Also deletes its data under data root (but not the source code)."
+    echo "       - \"indexed\" - mark the project(s) as indexed so it becomes visible in the UI"
```
Comment (tulinkry) [inline tools/Messages]: + info in README.txt

File: web/menu.jspf [hunk 9]
```diff
@@ -93,11 +93,15 @@ org.opensolaris.opengrok.web.Util"
...
         Math.min(15, projectsSize) %>"><%
         SortedSet<String> pRequested = cfg.getRequestedProjects();
-        for ( Group group : ph.getGroups() ) {
+        for (Group group : ph.getGroups()) {
             Set<Project> groupProjects = ph.getAllGrouped(group);
-            if ( groupProjects.size() > 0) {
+            if (groupProjects.size() > 0) {
                 %><optgroup label="<%= group.getName() %>"><%
                 for (Project p : groupProjects) {
+                    if (!p.isIndexed()) {
```
Comment (tulinkry) [inline web/menu.jspf]: There can be a situation when `groupProjects.size() > 0` this is true but with this `if` statement no project will be printed. Is that ok?

File: web/menu.jspf [hunk 10]
```diff
@@ -117,7 +121,11 @@ org.opensolaris.opengrok.web.Util"
         if (ph.getGroups().size() > 0 && ph.getAllUngrouped().size() > 0) {
             %><optgroup label="Other"><%
         }
-        for ( Project p : ph.getAllUngrouped()) {
+        for (Project p : ph.getAllUngrouped()) {
```
Comment (tulinkry) [inline web/menu.jspf]: same comment as above

**Comment by tulinkry** [comment]
> add a new project foo:
./tools/Messages -t foo -n project add

what if I search the project manually now after the first step (for example I have a wrong url containing the project name)?

what if I access the xref manually now? especially the root xref; is the project visible there now?

is there any problem with autorization of an existing yet not searchable project?

File: src/org/opensolaris/opengrok/configuration/RuntimeEnvironment.java [hunk 11]
```diff
@@ -1970,7 +2028,7 @@ public SuperIndexSearcher getIndexSearcher(String proj) throws IOException {
      * change so we go through the SearcherManager objects and close those where
      * the corresponding project is no longer present.
      */
-    private void refreshSearcherManagerMap() {
+    synchronized public void refreshSearcherManagerMap() {
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/configuration/RuntimeEnvironment.java]: is it necessary for this to be synchronized when the map is concurrent?

File: src/org/opensolaris/opengrok/configuration/messages/ProjectMessage.java [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+
+                // In case this project has just been incremetally indexed,
+                // its IndexSearcher needs a poke.
+                env.maybeRefreshIndexSearchers(getTags());
+
+                env.refreshDateForLastIndexRun();
+                break;
+        }
+
+        return null;
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/configuration/messages/ProjectMessage.java]: no "useful" output for the user? like "project abcd added" etc

File: src/org/opensolaris/opengrok/history/HistoryGuru.java [hunk 13]
```diff
@@ -620,6 +620,15 @@ public void createCache(Collection<String> repositories) {
         createCacheReal(getReposFromString(repositories));
     }
 
+    private HistoryCache getHistoryCache() {
+        HistoryCache cache = historyCache;
+        if (cache == null) {
+            cache = new FileHistoryCache();
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/history/HistoryGuru.java]: shouldn't this update also the `historyCache` variable?

File: src/org/opensolaris/opengrok/configuration/messages/ProjectMessage.java [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+import org.opensolaris.opengrok.logger.LoggerFactory;
+import org.opensolaris.opengrok.util.IOUtils;
+
+
+/**
+ * @author Vladimir Kotal
+ */
+public class ProjectMessage extends Message {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RuntimeEnvironment.class);
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/messages/ProjectMessage.java]: fixed

File: src/org/opensolaris/opengrok/history/FileHistoryCache.java (hunk 1 above)
Comment (vladak) [inline src/org/opensolaris/opengrok/history/FileHistoryCache.java]: yep

File: src/org/opensolaris/opengrok/history/FileHistoryCache.java (hunk 2 above)
Comment (vladak) [inline src/org/opensolaris/opengrok/history/FileHistoryCache.java]: I wanted to use the dirname either in the tests or elsewhere (which is bad idea really as everything from within `IndexDatabase` et al. should go through `HistoryGuru` and not directly to `FileHistoryCache`) so made a public duplicate. Removed the public object.

Comment (vladak) [inline src/org/opensolaris/opengrok/history/FileHistoryCache.java]: I guess the `FileHistoryCacheTest` can live with hardcoded `"historycache"`. If the name is changed the test will fail anyway.

File: src/org/opensolaris/opengrok/history/HistoryGuru.java (hunk 3 above)
Comment (vladak) [inline src/org/opensolaris/opengrok/history/HistoryGuru.java]: added

File: src/org/opensolaris/opengrok/history/HistoryGuru.java (hunk 4 above)
Comment (vladak) [inline src/org/opensolaris/opengrok/history/HistoryGuru.java]: fixed

File: src/org/opensolaris/opengrok/configuration/RuntimeEnvironment.java (hunk 5 above)
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/RuntimeEnvironment.java]: yes and yes :-)

File: test/org/opensolaris/opengrok/configuration/messages/ProjectMessageTest.java (hunk 6 above)
Comment (vladak) [inline test/org/opensolaris/opengrok/configuration/messages/ProjectMessageTest.java]: could not figure out how to add multiple conditions - the test file needs RepositoryInstalled.MercurialInstalled.class, RepositoryInstalled.SubversionInstalled.class, RepositoryInstalled.CvsInstalled.class

File: test/org/opensolaris/opengrok/configuration/messages/ProjectMessageTest.java (hunk 7 above)
Comment (vladak) [inline test/org/opensolaris/opengrok/configuration/messages/ProjectMessageTest.java]: fixed

File: tools/Messages (hunk 8 above)
Comment (vladak) [inline tools/Messages]: added

File: web/menu.jspf (hunk 9 above)
Comment (vladak) [inline web/menu.jspf]: I think so. Initially the menu listing can be empty and only group names will be displayed. I think this is better than suppressing empty groups. Alternatively, some count (indexed/not indexed) could be displayed.

File: web/menu.jspf (hunk 10 above)
Comment (vladak) [inline web/menu.jspf]: I wondered whether changing `ProjectHelper#cacheProjects()` to suppress the projects would be good idea (to make the counts on the index page next to the groups reflecting only indexed projects) however changing the caching might have undesired effects.

File: src/org/opensolaris/opengrok/configuration/RuntimeEnvironment.java (hunk 11 above)
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/RuntimeEnvironment.java]: should not be necessary (in the worst case the iterator will not reflect current state), removed.

File: src/org/opensolaris/opengrok/configuration/messages/ProjectMessage.java (hunk 12 above)
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/messages/ProjectMessage.java]: fixed

File: src/org/opensolaris/opengrok/history/HistoryGuru.java (hunk 13 above)
Comment (vladak) [inline src/org/opensolaris/opengrok/history/HistoryGuru.java]: removed this piece of code, will submit separate PR for #767

**Comment by vladak** [comment]
If the project is searched after it's added and before indexed the search will result in the `Failed to initialize search. Check the index.` error.

directory listings and xrefs will be displayed normally as they are generated on demand by default.

As for the authorization I think the indexed property of a project should be checked only after authorization which it currently is, at least in menu.jspf.

File: src/org/opensolaris/opengrok/configuration/RuntimeEnvironment.java [hunk 15]
```diff
@@ -130,6 +133,8 @@
     private int messagesInTheSystem = 0;
 
     private Statistics statistics = new Statistics();
+    
+    private static IndexTimestamp indexTime = new IndexTimestamp();
```
Comment (tarzanek) [inline src/org/opensolaris/opengrok/configuration/RuntimeEnvironment.java]: the truth is, that every project has its own index, so this should be a per project variable, but this is a good step forward, so let's start with this

**Comment by tarzanek** [comment]
lgtm, please merge, let's start testing this in real life

File: test/org/opensolaris/opengrok/configuration/messages/ProjectMessageTest.java (hunk 6 above)
Comment (tulinkry) [inline test/org/opensolaris/opengrok/configuration/messages/ProjectMessageTest.java]: Simple way is to somehow trick it in `setUp` using Assume.assumeTrue which causes the test to be ignored when the condition is not met

The nice but more complicated is to implement the ConditionalRun annotation to be repeatable
https://docs.oracle.com/javase/tutorial/java/annotations/repeating.html may be a starter

File: src/org/opensolaris/opengrok/configuration/RuntimeEnvironment.java (hunk 15 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/configuration/RuntimeEnvironment.java]: this is to be approached with #1390

Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/RuntimeEnvironment.java]: The question is how these per-project timestamps should be reflected in the UI.

**Comment by tulinkry** [comment]
> If the project is searched after it's added and before indexed the search will result in the Failed to initialize search. Check the index. error.

There should be something more intelligent - skip the project. Create an issue for this?

> Directory listings and xrefs will be displayed normally as they are generated on demand by default.

I wouldn't allow this as the not indexed project means "not ready for use". Could you create an issue to disable this for not indexed projects?

File: src/org/opensolaris/opengrok/history/FileHistoryCache.java (hunk 2 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/history/FileHistoryCache.java]: Yes the test can live without it. Or if you make the property protected it should be accessible from the test as it is the same package.

Doesn't NB hint you with naming conventions for constants?

File: web/menu.jspf (hunk 10 above)
Comment (tulinkry) [inline web/menu.jspf]: Yes this is exactly the place where the change should be made if necessary. But as you said if empty groups look fine to you just keep it.

For me it would be quite confusing why some groups arent displayed (unauthorized content - empty groups; no projects in the group) and some are (empty group but contains not indexed projects).

File: test/org/opensolaris/opengrok/configuration/messages/ProjectMessageTest.java (hunk 6 above)
Comment (vladak) [inline test/org/opensolaris/opengrok/configuration/messages/ProjectMessageTest.java]: hacked this around by using Assume and filed #1669 for the repeatable annotation.

File: web/menu.jspf (hunk 10 above)
Comment (vladak) [inline web/menu.jspf]: The latest change I did was to modify `cacheProjects()` to filter not-indexed projects so the changes in the cycle are probably not needed anymore.

**Comment by vladak** [comment]
>> If the project is searched after it's added and before indexed the search will result in the Failed to initialize search. Check the index. error.

>There should be something more intelligent - skip the project (current behaviour when the project does not exist). Create an issue for this?

Added a proper message to `searchHelper#prepareExec()`. The method receives the list already filtered by authorization framework.

>> Directory listings and xrefs will be displayed normally as they are generated on demand by default.

>I wouldn't allow this as the not indexed project means "not ready for use". Could you create an issue to disable this for not indexed projects?

I think that's fine as long as this functionality works (aka best effort) and gives proper error message when trying search.

**Comment by tulinkry** [comment]
> >> If the project is searched after it's added and before indexed the search will result in the Failed to initialize search. Check the index. error.
> 
> >There should be something more intelligent - skip the project (current behaviour when the project does not exist). Create an issue for this?
> 
> Added a proper message to `searchHelper#prepareExec()`. The method receives the list already filtered by authorization framework.

Ok

> >> Directory listings and xrefs will be displayed normally as they are generated on demand by default.
> 
> >I wouldn't allow this as the not indexed project means "not ready for use". Could you create an issue to disable this for not indexed projects?
> 
> I think that's fine as long as this functionality works (aka best effort) and gives proper error message when trying search.

But what for the root xref? The project will be available for browsing but not for searching that seems inconsistent. But still why not.

**Comment by vladak** [comment]
I realized (seeing that test case `PageConfigTest.java#testGetResourceFileList()` failed) that the `/xref` filtration is done via the changes done `cacheProjects()` so this should address the last concern.

Resynced with upstream master, all tests are passing now.
